File: dev/src/samples/NetSamples/Samples/Surfaces/SamplingPoints.cs
using System;

using Autodesk.AutoCAD.DatabaseServices;
using Autodesk.AutoCAD.EditorInput;
using Autodesk.AutoCAD.Geometry;
using Autodesk.AutoCAD.Runtime;
using Autodesk.Civil.DatabaseServices;

using CivilSurface = Autodesk.Civil.DatabaseServices.Surface;

[assembly: CommandClass(
    typeof(Autodesk.CivilizedDevelopment.SamplingPointsCommands))]

namespace Autodesk.CivilizedDevelopment
{
    public class SamplingPointsCommands : SimpleDrawingCommand
    {
        [CommandMethod("CDS_SamplePointsOnSurface")]
        public void CDS_SamplePointsOnSurface()
        {
            using (Transaction tr = startTransaction())
            {
                ITerrainSurface surface = getSurface() as ITerrainSurface;
                if (surface == null) return;
                ObjectId curveId = getCurveId();

                displayPoints(surface.SampleElevations(curveId));
            }
        }

        private CivilSurface getSurface()
        {
            ObjectId surfaceId = promptForSurface();
            if (surfaceId.IsNull)
            {
                return null;
            }
            CivilSurface surface = surfaceId.GetObject(OpenMode.ForRead)
                as CivilSurface;
            return surface;
        }

        private ObjectId promptForSurface()
        {
            PromptEntityOptions options = new PromptEntityOptions(
                "\nSelect a TIN Surface: ");
            options.SetRejectMessage(
                "\nThe selected object is not a TIN Surface.");
            options.AddAllowedClass(typeof(CivilSurface), false);

            PromptEntityResult result = _editor.GetEntity(options);
            if (result.Status == PromptStatus.OK)
            {
                // Everything is cool; we return the selected
                // surface ObjectId.
                return result.ObjectId;
            }
            return ObjectId.Null;   // Indicating error.
        }

        private ObjectId getCurveId()
        {
            PromptEntityOptions options = new PromptEntityOptions(
                "\nSelect entity: ");
            options.SetRejectMessage(
                "\nThe selected entity is not of the valid type.");
            PromptEntityResult result = _editor.GetEntity(options);
            if (result.Status == PromptStatus.OK)
            {
                return result.ObjectId;
            }
            return ObjectId.Null;
        }

        private void displayPoints(Point3dCollection points)
        {
            foreach (Point3d point in points)
            {
                _editor.WriteMessage("\n" + point.ToString());
            }
        }
    }
}

File: dev/src/samples/NetSamplesVB/Samples/Surfaces/SamplingPoints.vb
Imports System

Imports Autodesk.AutoCAD.DatabaseServices
Imports Autodesk.AutoCAD.EditorInput
Imports Autodesk.AutoCAD.Geometry
Imports Autodesk.AutoCAD.Runtime
Imports Autodesk.Civil.DatabaseServices

Imports CivilSurface = Autodesk.Civil.DatabaseServices.Surface

<Assembly: CommandClass(
GetType(Autodesk.CivilizedDevelopment.SamplingPointsCommands))> 

Namespace Autodesk.CivilizedDevelopment
  Public Class SamplingPointsCommands
    Inherits SimpleDrawingCommand
    <CommandMethod("CDS_SamplePointsOnSurface")> _
    Public Sub CDS_SamplePointsOnSurface()
      Using tr As Transaction = startTransaction()
        Dim surface As ITerrainSurface = TryCast(getSurface(), 
          ITerrainSurface)
        If surface Is Nothing Then
          Return
        End If
        Dim curveId As ObjectId = getCurveId()

        displayPoints(surface.SampleElevations(curveId))
      End Using
    End Sub



    Private Function getSurface() As CivilSurface
      Dim surfaceId As ObjectId = promptForSurface()
      If surfaceId.IsNull Then
        Return Nothing
      End If
      Dim surface As CivilSurface =
        TryCast(surfaceId.GetObject(OpenMode.ForRead), CivilSurface)
      Return surface
    End Function



    Private Function promptForSurface() As ObjectId
      Dim options As New PromptEntityOptions(
        vbLf & "Select a TIN Surface: ")
      options.SetRejectMessage(
        vbLf & "The selected object is not a TIN Surface.")
      options.AddAllowedClass(GetType(CivilSurface), False)

      Dim result As PromptEntityResult = _editor.GetEntity(options)
      If result.Status = PromptStatus.OK Then
        ' Everything is cool; we return the selected
        ' surface ObjectId.
        Return result.ObjectId
      End If
      Return ObjectId.Null
      ' Indicating error.
    End Function



    Private Function getCurveId() As ObjectId
      Dim options As New PromptEntityOptions(vbLf & "Select entity: ")
      options.SetRejectMessage(
        vbLf & "The selected entity is not of the valid type.")
      Dim result As PromptEntityResult = _editor.GetEntity(options)
      If result.Status = PromptStatus.OK Then
        Return result.ObjectId
      End If
      Return ObjectId.Null
    End Function



    Private Sub displayPoints(points As Point3dCollection)
      For Each point As Point3d In points
        _editor.WriteMessage(vbLf + point.ToString())
      Next
    End Sub


  End Class
End Namespace